mm2src/coins/lightning.rs

```diff
@@ -473,37 +621,512 @@ pub async fn open_channel(ctx: MmArc, req: OpenChannelRequest) -> OpenChannelRes
+            is_public: details.is_public,
+        }
+    }
+}
+
+#[derive(Serialize)]
+pub struct ListChannelsResponse {
+    channels: Vec<ChannelDetailsForRPC>,
+}
+
+pub async fn list_channels(ctx: MmArc, req: ListChannelsRequest) -> ListChannelsResult<ListChannelsResponse> {
+    let coin = lp_coinfind_or_err(&ctx, &req.coin).await?;
+    let ln_coin = match coin {
+        MmCoinEnum::LightningCoin(c) => c,
+        _ => return MmError::err(ListChannelsError::UnsupportedCoin(coin.ticker().to_string())),
     };
-    save_last_request_id_to_file(&last_request_id_path(&ctx, ticker), request_id)?;
-
-    let temporary_channel_id = open_ln_channel(
-        node_pubkey,
-        unsigned.outputs[0].value,
-        request_id,
-        req.announce_channel,
-        ln_coin.channel_manager.clone(),
+    let channels = ln_coin
+        .channel_manager
+        .list_channels()
+        .into_iter()
+        .map(From::from)
+        .collect();
+
+    Ok(ListChannelsResponse { channels })
+}
+
+#[derive(Deserialize)]
+pub struct GetChannelDetailsRequest {
+    pub coin: String,
+    pub channel_id: String,
+}
+
+#[derive(Serialize)]
+pub struct GetChannelDetailsResponse {
+    channel_details: ChannelDetailsForRPC,
+}
+
+pub async fn get_channel_details(
+    ctx: MmArc,
+    req: GetChannelDetailsRequest,
+) -> GetChannelDetailsResult<GetChannelDetailsResponse> {
+    let coin = lp_coinfind_or_err(&ctx, &req.coin).await?;
+    let ln_coin = match coin {
+        MmCoinEnum::LightningCoin(c) => c,
+        _ => return MmError::err(GetChannelDetailsError::UnsupportedCoin(coin.ticker().to_string())),
+    };
+    let channel_id =
+        hex::decode(req.channel_id.clone()).map_to_mm(|e| GetChannelDetailsError::DecodeError(format!("{}", e)))?;
+    let channel_details = ln_coin
+        .channel_manager
+        .list_channels()
+        .into_iter()
+        .find(|chan| chan.channel_id == channel_id[..])
+        .ok_or(GetChannelDetailsError::NoSuchChannel(req.channel_id))?
+        .into();
+
+    Ok(GetChannelDetailsResponse { channel_details })
+}
+
+#[derive(Deserialize)]
+pub struct GenerateInvoiceRequest {
+    pub coin: String,
+    pub amount_in_msat: Option<u64>,
+    pub description: String,
+}
+
+#[derive(Serialize)]
+pub struct GenerateInvoiceResponse {
+    invoice: String,
+    payment_hash: String,
+}
+
+/// Generates an invoice (request for payment) that can be paid on the lightning network by another node using send_payment.
+pub async fn generate_invoice(
+    ctx: MmArc,
+    req: GenerateInvoiceRequest,
+) -> GenerateInvoiceResult<GenerateInvoiceResponse> {
+    let coin = lp_coinfind_or_err(&ctx, &req.coin).await?;
+    let ln_coin = match coin {
+        MmCoinEnum::LightningCoin(c) => c,
+        _ => return MmError::err(GenerateInvoiceError::UnsupportedCoin(coin.ticker().to_string())),
+    };
+    let nodes_addresses = ln_coin.nodes_addresses.lock().clone();
+    for (node_pubkey, node_addr) in nodes_addresses {
```

Comment:
Do we really need to connect to all known nodes? Can we have a significant number here? I guess `generate_invoice` can be also stuck when there will be an attempt of connection to a malfunctioning node (bad connection/ping/etc.).

Reply:
node_addresses contains only channels node addresses https://github.com/KomodoPlatform/atomicDEX-API/blob/357cc30f3a45efa704c83746f7df86bf3573168a/mm2src/coins/lightning/ln_utils.rs#L434-L442
We have to make sure that we are connected to all the available channels counterparties when generating an invoice to [make these channels usable and add all the possible routing hints to the invoice](https://github.com/lightningdevkit/rust-lightning/blob/e43cfe135a5b68620cd51b9b982362c272b4ef77/lightning-invoice/src/utils.rs#L162-L184). This makes it easier for the payer to choose the best route based on the payer's channels configurations.

Reply:
Does it make sense to rename this field to something like `open_channel_nodes`?

Reply:
That's a more suitable name for sure, will rename it as part of the fixes in the next PR.